### DISCUSSION TITLE
Merge external subtitles into video

### DISCUSSION THREAD

**Discussion by damiensgit** [ptr727/PlexCleaner#227] [General]
I use Bazarr to download external subtitles but my LG C9 tv generally refuses to see external subtitles on my plex DNLA server so I have to run yet another tool to merge these external subs into the video file for better playback compatibility. If the video file only contains PGS subs it often causes plex to always transcode.

Looking at the docs it doesn't seem that it supports doing this so was wondering if this feature could be added?

For reference I'm using the following script to merge - as plex cleaner does a merge pass it would be cool if it could also do something similar and save multiple passes of the same file?
https://github.com/townsste/docker-mergesrt/blob/master/mergesrt.sh

**Comment by ptr727** [comment]
Plex server will seamlessly interleave and stream internal or external subtitles to the Plex client, i.e. the client does not know or care if the subs are internal or external. But transcoding or direct-stream/play will depend on the clients direct-stream compatibility and capability of the content type.

A quick google search says that the Plex WebOS client on LG does not support PGS subtitles (and other content types), and the server will thus transcode the video and burn the subs in server-side, if either the .

So, I'm wondering if what you are asking for is really "external subtitle file merging", or the "removal of [non-direct-stream-capable](https://support.plex.tv/articles/200471133-adding-local-subtitles-to-your-media/#:~:text=Other%20formats%20such%20as%20VOBSUB,the%20UTF%2D8%20character%20encoding.) subtitles"?

**Reply by damiensgit** [reply]
Thanks for the reply @ptr727 

To provide further clarity and problem case I'm trying to solve; whilst plex is supposed to do as you state, it sadly doesn't (the issue is a mix between plex/tv). E.g. If I have a movie file with just PGS subtitles (very common) which as you state isn't supported on my tv, and one single external SRT file in the correct name/format (plex does show the subtitle correctly in the web player), the LG TV player does not see any subtitles in this case. In addition, using the webos plex app on the tv (which is awful - very buggy and slow), the plex app will 99% of the time attempt to trans-encode to the tv as it has determined this the best solution but it most often causes playback issues. I have found an easy workaround; If I embed the external SRT subtitle into the movie, the default LG media player app correctly sees the subtitle, and the movie plays 100% fine with direct play and plex doesn't do any form of conversion. This is my ask, I would like to be able to include external subs I've downloaded into media as these seems to work well. Subtitles + 4k HDR + some audio formats really cause so much headache and it's so hard to tell when plex will decide to transcode. FYI burning PGS should be a fine solution but as I said above plex really doesn't seem to play well when it does this. I have hardware encoding support on the plex server side but the plex app just doesnt seem to be able to play correctly what is sent back. It seems there are a tonne of plex bugs the team aren't getting to on various clients so I prefer to avoid them for now hence I found this awesome tool.

To be clear, I don't want to remove non-direct-steam subtitles, just add any external.

Also to hijack my own thread: if a movie file contains just a single TrueHD audio track, is it possible to get plexcleaner to add a second re-encoded aac track from this?

**Comment by ptr727** [comment]
Are you sure that muxing the external subtitle into the MKV is really what solves the problem?  
Imagine the MKV contains an EN PGS subtitle, and you have an external EN SRT subtitle, what does Plex do, I don't know?  
What happens if you mux the EN SRT subtitle into the MKV that already contains an EN PGS subtitle, what does MkvMerge do, what does Plex do when there are now two EN subtitle tracks, one PGS and one SRT, I don't know?

It still seems to me that the presence of non-direct-stream content is the root problem.

As for auto-creating additional audio tracks, not something likely to be implemented, the tools does cleanup, not content creation.

**Reply by damiensgit** [reply]
Yeah all fair points and I get where you are coming from. It's a frustrating problem and the most recommended solution is "throw money at it and buy (another) external player" but I've been down this road and it's messy too with another set of draw backs, so I'm kinda naively hoping to fix it purely via software. 

For the example cases I can tell you as that is my exact situation:

> What happens if you mux the EN SRT subtitle into the MKV that already contains an EN PGS subtitle, what does MkvMerge do, what does Plex do when there are now two EN subtitle tracks, one PGS and one SRT, I don't know?

It works fine; it shows both in clients that support selecting subtitles. For players that don't support PGS, Plex sends the SRT as expected.

![image](https://github.com/ptr727/PlexCleaner/assets/50471/d2bc081f-596e-4efd-90ba-d10cefd97822)

> Imagine the MKV contains an EN PGS subtitle, and you have an external EN SRT subtitle, what does Plex do, I don't know?

It worked the same when the SRT subtitle was external, but in that case it just displayed a "external" tag. In this case, for players that didnt support PGS, plex didn't seem to correctly tell the client that it had a SRT ready to go.

I'll think about it some more, play some more and see if there's a better alt. Thanks for the feedback

**Reply by ptr727** [reply]
Ok, sounds good.

Reason I'm asking to get to the real issue is that the solution should be generally useful to other people, and I have personally wondered about just stripping out raster subtitles as they tend to be more problematic. E.g. a feature that will remove subtitles by type, e.g. PGS and VOB.

Keep an eye on the default tags for tracks, I'm noticing an increase in content that has multiple default flagged tracks, there should really be just one default per language, I'm not sure how it affects Plex.

**Reply by C0nw0nk** [reply]
Hey @ptr727 I just read this i did remove the vobsubs from my media items.

https://github.com/C0nw0nk/RemoveSubtitles

`dvb_subtitle dvb_teletext dvd_subtitle hdmv_pgs_subtitle`

FFMPEG lists a bunch here https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/matroska.c#L67

Found a few others uncommon here https://gist.github.com/tab1293/97a87523ad54005d6c659de7f30cca97

But reading what you said about multiple default flagged be it either audio or subtitle tracks does plexcleaner do the job of ensuring there is only 1 sub or audio track with a default flag if there are multiple ?
